pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,232 @@
+		return nil
+	}
+	if m.shutdownGracePeriodRequested == 0 {
+		return nil
+	}
+
+	systemBus, err := getSystemDbus()
+	if err != nil {
+		return err
+	}
+	m.dbusCon = systemBus
+
+	currentInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+	if err != nil {
+		return err
+	}
+
+	// If the logind's InhibitDelayMaxUSec as configured in (logind.conf) is less than shutdownGracePeriodRequested, attempt to update the value to shutdownGracePeriodRequested.
+	if m.shutdownGracePeriodRequested > currentInhibitDelay {
+		err := m.dbusCon.OverrideInhibitDelay(m.shutdownGracePeriodRequested)
+		if err != nil {
+			return err
+		}
+
+		err = m.dbusCon.ReloadLogindConf()
+		if err != nil {
+			return err
+		}
+
+		// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
+		updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+		if err != nil {
+			return err
+		}
+
+		if updatedInhibitDelay != m.shutdownGracePeriodRequested {
+			return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
+		}
+	}
+
+	err = m.aquireInhibitLock()
+	if err != nil {
+		return err
+	}
+
+	events, err := m.dbusCon.MonitorShutdown()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case isShuttingDown := <-events:
+				klog.V(1).Infof("Shutdown manager detected new shutdown event, isNodeShuttingDownNow: %t", isShuttingDown)
+				m.nodeShuttingDownMutex.Lock()
+				m.nodeShuttingDownNow = isShuttingDown
+				m.nodeShuttingDownMutex.Unlock()
+				if isShuttingDown {
+					m.processShutdownEvent()
+				} else {
+					m.aquireInhibitLock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *Manager) aquireInhibitLock() error {
+	lock, err := m.dbusCon.InhibitShutdown()
+	if err != nil {
+		return err
+	}
+	m.inhibitLock = lock
+	return nil
+}
+
+// ShutdownStatus will return an error if the node is currently shutting down.
+func (m *Manager) ShutdownStatus() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+
+	m.nodeShuttingDownMutex.Lock()
+	defer m.nodeShuttingDownMutex.Unlock()
+
+	if m.nodeShuttingDownNow {
+		return fmt.Errorf("node is shutting down")
+	}
+	return nil
+}
+
+func (m *Manager) processShutdownEvent() error {
+	klog.V(1).Infof("Shutdown manager processing shutdown event")
+	activePods := m.getPods()
+
+	nonCriticalPodGracePeriod := m.shutdownGracePeriodRequested - m.shutdownGracePeriodCriticalPods
+
+	var wg sync.WaitGroup
+	wg.Add(len(activePods))
+	for _, pod := range activePods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+
+			var gracePeriodOverride int64
+			if kubelettypes.IsCriticalPod(pod) {
+				gracePeriodOverride = int64(m.shutdownGracePeriodCriticalPods.Seconds())
```

Comment:
so this may be `0` is duration was less than 1 second. Should this be checked on input so this config will be treated as invalid?

Reply:
hmm, can you explain, I don't follow. `shutdownGracePeriodCriticalPods` can be zero, but I don't understand what you mean by "less than 1 second"...

Reply:
`gracePeriodOverride` casts `float` number of seconds to `int64`. So if configuration (which is Duration) was configured with millisecond duration, this value will be `0`.

Reply:
ah I see what you mean, nice catch. I added some validation as you suggested to ` pkg/kubelet/apis/config/validation/validation.go` to catch this case. 

Basically `ShutdownGracePeriod` and `ShutdownGracePeriodCriticalPods` must either be zero or if they're >= 0 they should be at least one second.